pkg/process/checks/process.go

```diff
@@ -255,116 +255,56 @@ func createProcCtrMessages(
-			return &model.CollectorProc{
-				HostName:          cfg.HostName,
-				NetworkId:         networkID,
-				Info:              sysInfo,
-				Processes:         p,
-				Containers:        c,
-				GroupId:           groupID,
-				ContainerHostType: cfg.ContainerHostType,
-			}
-		},
-	)
-
-	msgs = append(msgs, procCtrMessages...)
 
+	collectorProcs, totalProcs, totalContainers := chunkProcessesAndContainers(procsByCtr, containers, cfg.MaxPerMessage, maxWeightPerMessage)
 	// fill in GroupSize for each CollectorProc and convert them to final messages
 	// also count containers and processes
-	messages := make([]model.MessageBody, 0, len(msgs))
-	for _, m := range msgs {
-		m.GroupSize = int32(len(msgs))
+	messages := make([]model.MessageBody, 0, len(collectorProcs))
+	for _, m := range collectorProcs {
+		m.GroupSize = int32(len(collectorProcs))
+		m.HostName = cfg.HostName
+		m.NetworkId = networkID
+		m.Info = sysInfo
+		m.GroupId = groupID
+		m.ContainerHostType = cfg.ContainerHostType
+
 		messages = append(messages, m)
-		totalProcs += len(m.Processes)
-		totalContainers += len(m.Containers)
 	}
 
+	log.Tracef("Created %d process messages", len(messages))
+
 	return messages, totalProcs, totalContainers
 }
 
-// packProcCtrMessages packs container processes into messages using the next-fit bin packing algorithm. The
-// container and its processes are placed into a CollectorProc up to the provided capacity. Some containers may have
-// more processes than the supplied capacity, for these they simply get packed into its own message.
-func packProcCtrMessages(
-	capacity int,
+func chunkProcessesAndContainers(
 	procsByCtr map[string][]*model.Process,
 	containers []*model.Container,
-	msgFn ctrProcMsgFactory,
-) []*model.CollectorProc {
-	var msgs []*model.CollectorProc
-	var ctrs []*model.Container
-	var ctrProcs []*model.Process
+	maxChunkSize int,
+	maxChunkWeight int,
+) ([]*model.CollectorProc, int, int) {
+	chunker := &collectorProcChunker{}
+
+	totalProcs := len(procsByCtr[emptyCtrID])
 
-	space := capacity
+	chunkProcessesBySizeAndWeight(procsByCtr[emptyCtrID], maxChunkSize, maxChunkWeight, chunker)
 
+	totalContainers := 0
 	for _, ctr := range containers {
 		procs := procsByCtr[ctr.Id]
-
-		if len(procs) > capacity {
-			// this container has more process then the msg capacity, so we send it separately
-			msgs = append(msgs, msgFn(procs, ctr))
+		if len(procs) == 0 {
+			// can happen if a process is skipped (e.g. disallowlisted)
 			continue
 		}
-
-		if len(procs) > space {
-			// there is not enough space to fit the next set of container processes, so complete the payload with
-			// the previous container processes and reset
-			msgs = append(msgs, msgFn(ctrProcs, ctrs...))
-			ctrs = nil
-			ctrProcs = nil
-			space = capacity
+		totalProcs += len(procs)
+		totalContainers++
+		chunker.container = ctr
+		chunker.idx = 0
+		if len(chunker.collectorProcs) > 1 {
+			chunker.idx = len(chunker.collectorProcs) - 1
```

Comment:
I wonder if we could encapsulate the `idx` initialization inside of the chunker. And maybe instead of setting
```
chunker.container = ctr
```
we could pass the container as a parameter in the `chunkProcessesBySizeAndWeight` function and set the chunker container inside of it.

Reply:
This is an interesting idea, let me see if I can make it work.